### PR TITLE
Update permissions diff computation

### PR DESCRIPTION
**Pull request by meain**
Previously we were using permission ids to check if two permissions
are equal. While we were aware the permission ids can repeat across
different folder hierarchies we were not aware that this was also the
case within a single folder tree if we delete and break the
inheritance chain.

Assume the following scenario:

1. Create folder `a` and `a/b`
2. Assign user1:read permission to `a`
3. Remove the inherited permission from `a/b`
4. Assign user1:write permission to `a/b`

In this case both the user1:read and user1:write will get the same permissions id.

---

#### Does this PR need a docs update or release note?

- [ ] :white_check_mark: Yes, it's included
- [ ] :clock1: Yes, but in a later PR
- [x] :no_entry: No

#### Type of change

- [ ] :sunflower: Feature
- [x] :bug: Bugfix
- [ ] :world_map: Documentation
- [ ] :robot: Supportability/Tests
- [ ] :computer: CI/Deployment
- [ ] :broom: Tech Debt/Cleanup

#### Issue(s)

* fixes https://github.com/alcionai/corso/issues/3116

#### Test Plan

- [ ] :muscle: Manual
- [x] :zap: Unit test
- [ ] :green_heart: E2E

### CODE REVIEW

File: src/internal/connector/onedrive/permission.go
Comment (ashmrtn) [inline src/internal/connector/onedrive/permission.go:133]: do we know the conditions under which the entityID or the email will be empty? Can we get comments on how that could happen please

Also, do we need to worry about checking equality across backup versions? For example, would it be possible for the child to have a struct from an older backup that used Email and the parent to be from a newer backup that used EntityID?

**Review by ashmrtn — COMMENTED** [review]
generally looks good. Few open questions that need answered before approval though

File: src/internal/connector/onedrive/permission.go
Comment (meain) [reply src/internal/connector/onedrive/permission.go:193]: It is sensitive to ordering. I would have liked to move this logic to the [function](https://github.com/alcionai/corso/blob/762f7bc69efa59ba1632b49516fa7d68f031801f/src/internal/connector/onedrive/permission.go#L355) that does the restore, but that function only has the diffed version. That said, the function that does the restore already relies on removing before adding. Let me add a comment on both though.

File: src/internal/connector/onedrive/permission.go
Comment (meain) [reply src/internal/connector/onedrive/permission.go:138]: I've added comments explaining these things in the code in https://github.com/alcionai/corso/pull/3117/commits/7e1ae0bc68153cbbb03c4e95948f77e1fe4d692f

File: src/internal/connector/onedrive/permission.go
Comment (meain) [reply src/internal/connector/onedrive/permission.go]: I did some more digging and realized that this can be a problem is the previous one has empty permissions.

Comment (meain) [reply src/internal/connector/onedrive/permission.go]: On second though, let me remove this completely for now. It is a much deeper rabbit hole with not as much importance IMO. I can't think of a straight forward way to do this as we might have to ideally track if individual permissions are inherited or custom for which data is not even available AFAIC. There is `inheritedFrom` property, but it never shows up for any permission entries that are for items with custom permissions. I've created another issue to track this at https://github.com/alcionai/corso/issues/3126

Let me also loop in @gmatev  in case he has any thoughts on this.

File: src/internal/tester/config.go
Comment (meain) [inline src/internal/tester/config.go:130]: I've added a new user to make sure that we have enough permissions combinations to test all the different scenarios. @ashmrtn will need you help to add another CI variable for this.

File: src/internal/connector/onedrive/permission.go
Comment (meain) [reply src/internal/connector/onedrive/permission.go]: We can always make it not inherit by deleting and recreating common ones. We were previously relying on ids to do this, but as we have seen, that is not reliable.

**Comment by ashmrtn** [comment]
Can you add a summary of what this PR aims to do please? I see that some of the tests have been bulked up, but others have also been disabled. This leaves me unsure what high-level thing I should be gauging the code against

**Comment by meain** [comment]
@ashmrtn 
> Can you add a summary of what this PR aims to doplease?

The problem we are trying to solve here is that we were assuming that permissions where inherited if the permissions on parent and child had the same id. This is not true.

In the below scenario, even thought the permissions is not inherited(or same), they end up with same id:

-   Create folder a and a/b
-   Assign user1:read permission to a
-   Remove the inherited permission from a/b
-   Assign user1:write permission to a/b

This creates problems with restoring permissions as we might end up with incorrect permissions. The fix here reverts to checking the content of the permissions. While this now ensures that we have the proper set of permissions on the final item, we cannot ensure that the inheritance is maintained. For example if `a` has `user1:read` and `a/b` has `user1:read` that was not inherited(deleted and recreated on the  item), after going through corso, it will be as if `a/b` inherited the permissions from `a`.

**Review by ashmrtn — APPROVED** [review]
so I'm still a little unclear on the underlying problem, but if the issue is ensuring we get unique permissions on the Microsoft backend so that changes don't affect things we don't mean to when changing other permission sets it may be worth looking into alternate ways to adjust permissions on items. It seems like things like [sending a sharing invite](https://learn.microsoft.com/en-us/graph/api/driveitem-invite?view=graph-rest-1.0&tabs=http) allow setting a flag that says not to inherit permissions. There's also [creating a sharing link](https://learn.microsoft.com/en-us/graph/api/driveitem-createlink?view=graph-rest-1.0&tabs=http) (afaik we don't want this but adding for completeness) which also has an inheritance flag and can be modified later on to [grant permissions](https://learn.microsoft.com/en-us/graph/api/permission-grant?view=graph-rest-1.0&tabs=http) to specific people with specific roles

Since both of the above have flags that control inheritance it seems like it would maybe cause Microsoft to give us unique permission IDs for them?

**Comment by meain** [comment]
> ... proper flag value for "do not inherit" should fix that though ...

Do not inherit on the item, not on an individual permission which is what `retainInheritedPermissions` does.

**Pull request merged**
